mindmeld/active_learning/data_loading.py

```diff
@@ -0,0 +1,479 @@
+        if training_level == TRAIN_LEVEL_DOMAIN:
+            return [f"{q.domain}" for q in queries]
+        elif training_level == TRAIN_LEVEL_INTENT:
+            return [f"{q.domain}|{q.intent}" for q in queries]
+        else:
+            raise ValueError(
+                f"Invalid label_type {training_level}. Must be '{TRAIN_LEVEL_DOMAIN}'"
+                f" or '{TRAIN_LEVEL_INTENT}'"
+            )
+
+
+class LogQueriesLoader:
+    def __init__(self, app_path: str, training_level: str, log_file_path: str):
+        """This class loads data as processed queries from a specified log file.
+        Args:
+            app_path (str): Path to the MindMeld application.
+            training_level (str): The hierarchy level to train ("domain" or "intent")
+            log_file_path (str): Path to the log file with log queries.
+        """
+        self.app_path = app_path
+        self.training_level = training_level
+        self.log_file_path = log_file_path
+
+    @staticmethod
+    def filter_raw_text_queries(log_queries_iter) -> List[str]:
+        """Removes duplicates in the text queries.
+
+        Args:
+            log_queries_iter (generator): Log queries generator.
+        Returns:
+            filtered_text_queries (List[str]): a List of filtered text queries.
+        """
+        return list(set(q for q in log_queries_iter))
+
+    def convert_text_queries_to_processed(
+        self, text_queries: List[str]
+    ) -> List[ProcessedQuery]:
+        """Converts text queries to processed queries using an annotator.
+
+        Args:
+            text_queries (List[str]): a List of text queries.
+        Returns:
+            queries (List[ProcessedQuery]): List of processed queries.
+        """
+        logger.info("Loading a Bootstrap Annotator to process log queries.")
+        annotator_params = DEFAULT_AUTO_ANNOTATOR_CONFIG
+        annotator_params["app_path"] = self.app_path
+        bootstrap_annotator = BootstrapAnnotator(**annotator_params)
+        return bootstrap_annotator.text_queries_to_processed_queries(
+            text_queries=text_queries
+        )
+
+    @property
+    def queries(self):
+        log_queries_iter = read_query_file(self.log_file_path)
+        filtered_text_queries = LogQueriesLoader.filter_raw_text_queries(
+            log_queries_iter
+        )
+        return self.convert_text_queries_to_processed(filtered_text_queries)
+
+    @property
+    def class_labels(self):
+        return QueryLoader.get_class_labels(self.training_level, self.queries)
+
+
+class DataBucket:
+    """Class to hold data throughout the Active Learning training pipeline.
+    Responsible for data conversion, filtration, and storage.
+    """
+
+    def __init__(
+        self,
+        label_map: LabelMap,
+        test_queries: List[ProcessedQuery],
+        unsampled_queries: List[ProcessedQuery],
+        sampled_queries: List[ProcessedQuery],
+        newly_sampled_queries: List[ProcessedQuery] = None,
```

Comment:
Its more performant to track the indices to ProcessedQuery objects than the actual objects since mem will blow up especially with `deepcopy` calls to databuckets.

Reply:
Now tracking indices instead. Using ProcessedQueryList objects and the new sqlite-backed QueryCache